README
This repo contains some c++ code.

Variable Storage durations in C++:
    Static storage - variable dies when the program exits.
    local storage - variables declared within a block will die once the execution comes out of the block (Variable lives within that block).
    Dynamic storage - We decide on when the variable dies.

Variable Scope in C++:
Defines where we can access a variable


Inline static member variables only work in c++ > 17

CHANGES.md
# 100 Days C++

## Day 13 : 
    - Assignment9 (Year Calendar)
    - InfiniteLoop (optimised calculator)
    - Arrays
    - exercise10 (Remove duplicates)
    - exercise10.0 (Only unique elements)

## Day 14 :
    - Size of an array
    - exercise11 (Check Sorted collection)
    - Character Array
    - exercise12 (Hunt for Vowels)
    - Random Numbers (With and without seed using current time)

## Day 15 :
    - Fortune teller
    - exercise13 (Find common elements between two arrays)
    - Multi Dimentional Arrays [int]
    - Multi Dimentional Arrays [char]
    - Optimizing Fortune Teller 

## Day 16 :
    - Assignment10 (The Calculator quiz)

## Day 17 :
    - pointers overview
    - Fortune Teller optimization using pointers
    - Const Pointer and pointer to const
    - Array of const pointers to const character
    - Exercise 14 (Where does the big guy live?)
    - Swap Arrays (Constant Time complexity using pointers)

## Day 18 :
    - Pointer Arithmetic
    - Exercise 15 (Shoot forward using pointers)
    - Distance between elements (pointers)
    - Exercise 16 (Finding Nemo - Address of a little guy using pointer arithmetic)
    - PointerArithmetic comparison (Compares the adresses)

## Day 19 :
    - Dynamic memory Allocation
    - Dangling Pointers
    - Memory Leaks

## Day 20 :
    - Dynamic Arrays
    - Exercise 17 (Merge two arrays dynamic)
    - References
    - Exercise 18 (Declare references)
    - References vs pointers
    - References to update values in a range based for loop(in-place updates)

## Day 21 :
    - Character manipulation
    - Exercise 19 (How many vowels and consonants)
    - Exercise 20 (Replace spaces with underscores)

## Day 22 :
    - C-strings

## Day 23 :
    - C strings (Concatenation and copy)
    - Exercise 21 (Get something meaningful out of gibberish character arrays)
    - std::string
    - Exercise 22 (Build from raw pieces)
    - Size and capacity of std::string
    - Modify std::strings
    - Compare std::strings
    - Exercise 23 (Greatest string in the array of strings)
    - String searching

## Day 24 :
    - c++ functions
    - Exercise 25 (Function to get sum of digits present in a number)
    - Function declaration and function definition

## Day 25 : 
    - Multiple files with headers (Linkers)
    - Exercise 27 (Check if a number is a palindrome)
    - Pass by value
    - Pass by const value
    - Pass by pointers
    - Pass by pointers to const
    - Pass by const pointer to a const value
    - Pass by reference
    - Pass by const reference
    - Exercise 28 (Find the characters)

## Day 26 :
    - Exercise 28 continue
    - Array function parameters

## Day 27 :
    - Sized Array function parameters
    - Pass array as reference
    - Multi dimentional array parameters
    - Default arguments
    - Implicit conversions
    - Implicit conversions with references
    - Implicit conversion with pointers
    - String Views
    - Convert from string_view to string
    - Constexpr function
    - Consteval function
    - Enum classes
    - Exercise 29 (Day of the week)
    - 'using' Keyword with enum
    - Legacy enumerators

## Day 28 :
    - Arguments to a main function
    - Calculator V1 using arguments to main function
    - IO parameters in functions
    - Exercise 30 (Check if the sum is even)
    - Return by value

## Day 29 :
    - Return by reference
    - Return by pointer
    - Return array element by pointer
    - Exercise 31 (Fixed instructor's messed up code)
    - Auto type-deductions
    - Function return type deduction
    - Function return type deduction for reference
    - Function overloading with different parameters
    - Function overloading with pointer parameters
    - Function overloading with reference parameters
    - Exercise 32 (Overloads)
    - Overloading with const parameter
    - Overloading pointer/const_pointer to const
    - Overloading const references as parameters

## Day 30 :
    - lambda functions overview
    - Capture lists capture specific values
    - Capture lists capture all values outside lambda
    - Misfits - static variables
    - Exercise 33 (Self tracking function)
    - Exercise 34 (Odd series)
    - Inline functions
    - recurssion
    - Exercise 35 (Factorial using recurrsion)

## Day 31 :
    - Function templates overview
    - Exercise 36 (Hunt the elements down using function templates)
    - Function template type deduction
    - Function templates by reference
    - EXercise 37 (Swap them up)
    - Template specialisation
    - Exercise 38 (find em const char*)
    - Function template with overloading
    - Exercise 39 (const char* function template overloads)

## Day 32 :
    - Function template with multiple parameters
    - Exercise 40 (Put them number together)
    - Function template return type deduction with auto
    - Function template return type deduction with Decltype and trailing return types
    - Decl type auto
    - DEfault arguments templates

## Day 33 :
    - Non-type template parameters
    - EXercise 41 (Find them non type template parameters)
    - Auto template types
    - Type traits
    - constexpr if
    - concepts
    - custom concepts
    - Assignment 11
    - Requires clause
    - Combining concepts
    - Assignment 12

## Day 34 :
    - Classes overview
    - Exercise 42 ( Class box)
    - Exercise 43 ( 2D )
    - Constructor
    - Exercise 44 (Constructors with box class)
    - Exercise 45 (Constructors with points class)
    - Defaulted constructors
    - Setters and getters
    - EXercise 46 (Point printing like a gentleman)
    - Exercise 47 (Classes in multiple files)

## Day 35 :
    - Arrow pointers
    - Destructors
    - Order of constructors and destructors
    - this pointer
    - Exercise 48 (this setter chains using pointers)
    - Exercise 49 (this setter chains using references)
    - struct vs class
    - size of an object
    - Const objects
    - const objects as function arguments

arrays.cpp
#include <iostream>
int main()
{

    // Declare an array of ints
    // int scores [10]; // Junk data

    // Read data

    // std:: cout << " scores [0] : " << scores[0] << std::endl;
    // std:: cout << " scores [1] : " << scores[1] << std::endl;

    // Read with a loop
    /*
    for( size_t i {0} ; i < 10 ; ++i){
        std::cout << "scores [" << i << "] : " << scores[i] << std::endl;
    }
    */

    /*
     scores[0] = 20;
     scores[1] = 21;
     scores[2] = 22;

     //Print the data out
     for( size_t i {0} ; i < 10 ; ++i){
         std::cout << "scores [" << i << "] : " << scores[i] << std::endl;
     }
     */

    // Write data in a loop
    /*
    for(size_t i{0}; i < 10 ; ++i){
        scores[i] = i * 10;
    }

     //Print the data out
     for( size_t i {0} ; i < 10 ; ++i){
         std::cout << "scores [" << i << "] : " << scores[i] << std::endl;
     }
     */

    // Declare and initialize at the same time
    /*
    double salaries[5] {12.7, 7.5, 13.2, 8.1, 9.3};

    for(size_t i{0}; i < 5; ++i){
        std::cout << "salary[" << i << "] : " << salaries[i] << std::endl;
    }
    */

    // If you don't initialize all the elements, those you leave out
    // are initialized to 0
    /*
    int families[5] {12, 7, 5};

    for(size_t i{0}; i < 5; ++i){
        std::cout << "families[" << i << "] : " << families[i] << std::endl;
    }
    */

    // Omit the size of the array at declaration
    /*
     int class_sizes[] {10,12,15,11,18,17,23,56};

     // Will print this with a range based for loop
     for(auto value : class_sizes){
         std::cout << "value : " << value << std::endl;
     }
     */

    // Read only arrays
    /*
    const int birds[] {10,12,15,11,18,17,23,56};
    birds[2] = 8;
    */

    // Sum up scores array, store result in sum

    int scores[]{2, 5, 8, 2, 5, 6, 9};
    int sum{0};

    for (int element : scores)
    {
        sum += element;
    }
    std::cout << "Score sum : " << sum << std::endl;

    // Size of an array

    std::cout << "Size of array is : " << sizeof(scores) << std::endl;
    // Standard way to get number of elements in array

    std::cout << "Number of elements in the array is : " << sizeof(scores) / sizeof(scores[0]) << std::endl;

    // Easy way to get number of elements in an array

    std::cout << "Number of elements in array easy way is : " << std::size(scores) << std::endl;

    return 0;
}
